File: java/src/main/java/recipes/FundingRates.java
// Compare perpetual-swap funding rates across derivatives exchanges.
package recipes;

import io.github.ccxt.exchanges.Binance;
import io.github.ccxt.exchanges.Bybit;
import io.github.ccxt.exchanges.Okx;
import io.github.ccxt.types.FundingRate;

public class FundingRates {

    public static void main(String[] args) {
        String symbol = args.length > 0 ? args[0] : "BTC/USDT:USDT";
        print("binance", new Binance().fetchFundingRate(symbol));
        print("bybit", new Bybit().fetchFundingRate(symbol));
        print("okx", new Okx().fetchFundingRate(symbol));
    }

    static void print(String id, FundingRate rate) {
        String pct = (rate.fundingRate != null) ? String.format("%.4f%%", rate.fundingRate * 100) : "n/a";
        String next = (rate.fundingDatetime != null) ? rate.fundingDatetime : "n/a";
        System.out.printf("%-8s funding %s  next %s%n", id, pct, next);
    }
}

File: csharp/recipes/FundingRates.cs
// Compare perpetual-swap funding rates across derivatives exchanges.
using ccxt;

namespace Recipes;

public static class FundingRates
{
    public static async Task Run(string[] args)
    {
        var symbol = args.Length > 0 ? args[0] : "BTC/USDT:USDT";
        await Print("binance", new Binance().FetchFundingRate(symbol));
        await Print("bybit", new Bybit().FetchFundingRate(symbol));
        await Print("okx", new Okx().FetchFundingRate(symbol));
    }

    private static async Task Print(string id, Task<FundingRate> request)
    {
        try
        {
            var rate = await request;
            var pct = (rate.fundingRate != null) ? $"{rate.fundingRate * 100:F4}%" : "n/a";
            Console.WriteLine($"{id,-8} funding {pct}  next {rate.fundingDatetime ?? "n/a"}");
        }
        catch (Exception e)
        {
            Console.WriteLine($"{id,-8} error: {e.Message}");
        }
    }
}
